File: Linked List/1D LL/Java/Singly_Linked_List.java
public class Singly_Linked_List{
    public static void main(String[] args){
        Linked_List list = new Linked_List();
        list.insert(1);
        list.insert(2);
        list.insert(3);
        list.insert(4);
        list.insert(5);

        System.out.println("Original List:");
        list.display();

        list.delete(3);
        System.out.println("After Deleting 3:");
        list.display();

        list.insertAtBeginning(0);
        System.out.println("After Inserting 0 at Beginning:");
        list.display();
    }
}

class Linked_List{
    Node head;
    // Insert new node at the end
    public void insert(int data){
        Node newNode = new Node(data);

        if(head == null){
            head = newNode;
            return;
        }

        Node current = head;
        while (current.next != null) { 
            current = current.next;
        }

        current.next = newNode;
    }

    //Insert at the beginning
    public void insertAtBeginning(int data){
        Node newNode = new Node(data);

        newNode.next = head;
        head = newNode;       
    }

    // Delete a Node
    public void delete(int data){
        if(head == null) return;

        if(head.data == data){
            head = head.next;
            return;
        }

        Node current = head;
        while (current.next != null) { 
            if(current.next.data == data){
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }
    }

    // Search for a value
    public boolean search(int data){
        Node current = head;

        while (current != null) { 
            if(current.data == data) return true;
            current = current.next;
        }
        return false;
    }

    // Display the Linked List
    public void display(){
        Node current = head;

        while (current != null) { 
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }
}

class Node{
    int data;
    Node next;

    Node(int data){
        this.data = data;
        this.next = null;
    }
}

File: Linked List/1D LL/Python/singly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None

    # 1. Append (Add to the end) - O(n)
    def append(self, data):
        new_node = Node(data)

        if not self.head:
            self.head = new_node
            return

        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    # 2. Prepend (Add to the beginning) - O(1)
    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    # 3. Delete a node by value - O(n)
    def delete(self, key):
        current = self.head

        # Case 1: Key is at the head
        if current and current.data == key:
            self.head = current.next
            current = None
            return

        # Case 2: Key is somewhere in the middle or end
        prev = None
        while current and current.data != key:
            prev = current
            current = current.next

        if not current:
            print(f"Value '{key}' not found in list.")
            return

        prev.next = current.next
        current = None

    # 4. Traversal / Printing - O(n)
    def display(self):
        nodes = []
        current = self.head
        while current:
            nodes.append(str(current.data))
            current = current.next
        print(" -> ".join(nodes) if nodes else "Empty List")


ll = SinglyLinkedList()

ll.append(10)
ll.append(20)
ll.append(30)
ll.display()  # Output: 10 -> 20 -> 30

ll.prepend(5)
ll.display()  # Output: 5 -> 10 -> 20 -> 30

ll.delete(20)
ll.display()  # Output: 5 -> 10 -> 30